obstore/python/obstore/auth/earthdata.py

```diff
@@ -6,157 +6,419 @@
-            "secret_access_key": creds["secretAccessKey"],
-            "token": creds["sessionToken"],
-            "expires_at": datetime.fromisoformat(creds["expiration"]),
-        }
+        if self._session is None:
+            msg = "credential provider was closed"
+            raise ValueError(msg)
+
+        metadata = self._get_credentials(self._session)
+
+        return _metadata_to_s3_credential(metadata)
 
     def close(self) -> None:
         """Close the underlying session.
 
         You should call this method after you've finished all obstore calls to close the
-        underlying [requests.Session][].
+        underlying [requests.Session][].  If you supplied your own session object,
+        it will not be closed.
         """
-        self.session.close()
+        if self._close:
+            self._session = None
+            self._close()
+
+
+def _make_get_credentials(
+    credentials_url: str,
+    host: str,
+    auth: str | tuple[str, str] | None,
+) -> Callable[[requests.Session], Mapping[str, str]]:
+    if isinstance(auth, str):
+        return lambda session: _get_with_token(session, credentials_url, auth)
+    return lambda session: _get_with_basic_auth(session, credentials_url, host, auth)
+
+
+def _get_with_token(
+    session: requests.Session,
+    credentials_url: str,
+    token: str,
+) -> Mapping[str, str]:
+    with session.get(
+        credentials_url,
+        allow_redirects=False,
+        headers={"Authorization": f"Bearer {token}"},
+    ) as r:
+        r.raise_for_status()
+
+        if r.is_redirect:
+            # We were redirected; basic auth creds are invalid or not found via netrc
+            r.status_code = 401
+            r.reason = "Unauthorized"
+            r.raise_for_status()
+
+        return r.json()
+
+
+def _get_with_basic_auth(
+    session: requests.Session,
+    credentials_url: str,
+    host: str,
+    basic_auth: tuple[str, str] | None,
+) -> Mapping[str, str]:
+    with session.get(credentials_url, allow_redirects=False) as r:
+        r.raise_for_status()
+        location = r.headers["location"]
+
+    # We were redirected, so we must use basic auth credentials with the
+    # redirect location.  If the host of the redirect is the same host we have
+    # creds for, pass them along; otherwise, netrc will be used (implicitly),
+    # if the session's trust_env attribute is set to True (default).
+
+    redirect_host = str(urlparse(location).hostname)
+    auth = basic_auth if redirect_host == host else None
+
+    with session.get(location, auth=auth) as r:
+        r.raise_for_status()
+
+        try:
+            return r.json()
+        except json.JSONDecodeError:
+            # Content is not JSON; basic auth creds are invalid or not found via netrc
+            r.status_code = 401
+            r.reason = "Unauthorized"
+            r.raise_for_status()
+            return {}  # Never reached, but necessary for type checking
```

Comment:
Nit: I usually prefer an `assert False` to make clear the code is unreachable. I think that also makes type checking pass?

Reply:
Ah, sure, let me see.